File: src/lib/metrics.ts
import type { ActivityRecord, ActivityType } from "../types.js";
import { ACTIVITY_TYPES } from "../types.js";

export interface PrecisionRecallF1 {
  precision: number;
  recall: number;
  f1: number;
}

export interface PerClassMetric extends PrecisionRecallF1 {
  activityType: ActivityType;
  support: number;
}

export interface Pair {
  predicted: ActivityRecord;
  label: ActivityRecord;
}

function safeDiv(numerator: number, denominator: number): number {
  return denominator === 0 ? 0 : numerator / denominator;
}

export function normalizeOfficeName(name: string): string {
  return name.trim().toLowerCase().replace(/\s+/g, " ").replace(/[.,]/g, "");
}

export function officeNameAccuracy(pairs: readonly Pair[]): number {
  if (pairs.length === 0) return 0;
  const correct = pairs.filter(
    (p) =>
      normalizeOfficeName(p.predicted.office_name) ===
      normalizeOfficeName(p.label.office_name),
  ).length;
  return correct / pairs.length;
}

export function activityTypeAccuracy(pairs: readonly Pair[]): number {
  if (pairs.length === 0) return 0;
  const correct = pairs.filter(
    (p) => p.predicted.activity_type === p.label.activity_type,
  ).length;
  return correct / pairs.length;
}

export function perClassActivityTypeMetrics(
  pairs: readonly Pair[],
): PerClassMetric[] {
  return ACTIVITY_TYPES.map((activityType) => {
    let tp = 0;
    let fp = 0;
    let fn = 0;
    let support = 0;
    for (const p of pairs) {
      const predicted = p.predicted.activity_type === activityType;
      const actual = p.label.activity_type === activityType;
      if (actual) support += 1;
      if (predicted && actual) tp += 1;
      else if (predicted && !actual) fp += 1;
      else if (!predicted && actual) fn += 1;
    }
    const precision = safeDiv(tp, tp + fp);
    const recall = safeDiv(tp, tp + fn);
    const f1 = safeDiv(2 * precision * recall, precision + recall);
    return { activityType, precision, recall, f1, support };
  });
}

export function nextActionDateExactMatchRate(pairs: readonly Pair[]): number {
  if (pairs.length === 0) return 0;
  const matches = pairs.filter(
    (p) => p.predicted.next_action_date === p.label.next_action_date,
  ).length;
  return matches / pairs.length;
}

/** Fraction of records where predicted-null-ness of next_action matches the label's. */
export function nextActionNullHandlingAccuracy(pairs: readonly Pair[]): number {
  if (pairs.length === 0) return 0;
  const matches = pairs.filter(
    (p) =>
      (p.predicted.next_action === null) === (p.label.next_action === null),
  ).length;
  return matches / pairs.length;
}

function normalizeSetItem(item: string): string {
  return item.trim().toLowerCase();
}

function setPrecisionRecallF1(
  predicted: string[],
  label: string[],
): PrecisionRecallF1 {
  const predSet = new Set(predicted.map(normalizeSetItem));
  const labelSet = new Set(label.map(normalizeSetItem));
  let tp = 0;
  for (const item of predSet) {
    if (labelSet.has(item)) tp += 1;
  }
  const precision = safeDiv(tp, predSet.size);
  const recall = safeDiv(tp, labelSet.size);
  const f1 = safeDiv(2 * precision * recall, precision + recall);
  return { precision, recall, f1 };
}

/** Averages set precision/recall/F1 for a list-valued field across all pairs. */
export function averageSetMetrics(
  pairs: readonly Pair[],
  field: "contacts" | "products_mentioned",
): PrecisionRecallF1 {
  if (pairs.length === 0) return { precision: 0, recall: 0, f1: 0 };
  const perPair = pairs.map((p) =>
    setPrecisionRecallF1(p.predicted[field], p.label[field]),
  );
  const avg = (key: keyof PrecisionRecallF1): number =>
    perPair.reduce((sum, m) => sum + m[key], 0) / perPair.length;
  return { precision: avg("precision"), recall: avg("recall"), f1: avg("f1") };
}

/**
 * Hallucination rate: fraction of records where the model filled in
 * next_action or next_action_date with a non-null value when the label says
 * that field should be null.
 */
export function hallucinationRate(pairs: readonly Pair[]): number {
  if (pairs.length === 0) return 0;
  const hallucinated = pairs.filter((p) => {
    const nextActionHallucinated =
      p.label.next_action === null && p.predicted.next_action !== null;
    const nextActionDateHallucinated =
      p.label.next_action_date === null &&
      p.predicted.next_action_date !== null;
    return nextActionHallucinated || nextActionDateHallucinated;
  }).length;
  return hallucinated / pairs.length;
}

File: python/src/fieldnotes/metrics.py
"""Exact port of src/lib/metrics.ts. Every function here must reproduce the
TypeScript version bit for bit: same operation order, same constants, plain
Python floats only (no numpy, no math.fsum, no rounding)."""

from __future__ import annotations

import re
from typing import Literal, TypedDict

from fieldnotes.types import ACTIVITY_TYPES, ActivityRecord, ActivityType

_WHITESPACE_RE = re.compile(r"\s+")
_PUNCTUATION_RE = re.compile(r"[.,]")

SetField = Literal["contacts", "products_mentioned"]


class PrecisionRecallF1(TypedDict):
    precision: float
    recall: float
    f1: float


class PerClassMetric(TypedDict):
    activityType: ActivityType
    precision: float
    recall: float
    f1: float
    support: int


class Pair(TypedDict):
    predicted: ActivityRecord
    label: ActivityRecord


def safe_div(numerator: float, denominator: float) -> float:
    return 0.0 if denominator == 0 else numerator / denominator


def normalize_office_name(name: str) -> str:
    stripped = name.strip().lower()
    collapsed = _WHITESPACE_RE.sub(" ", stripped)
    return _PUNCTUATION_RE.sub("", collapsed)


def office_name_accuracy(pairs: list[Pair]) -> float:
    if len(pairs) == 0:
        return 0.0
    correct = sum(
        1
        for p in pairs
        if normalize_office_name(p["predicted"]["office_name"])
        == normalize_office_name(p["label"]["office_name"])
    )
    return correct / len(pairs)


def activity_type_accuracy(pairs: list[Pair]) -> float:
    if len(pairs) == 0:
        return 0.0
    correct = sum(
        1 for p in pairs if p["predicted"]["activity_type"] == p["label"]["activity_type"]
    )
    return correct / len(pairs)


def per_class_activity_type_metrics(pairs: list[Pair]) -> list[PerClassMetric]:
    results: list[PerClassMetric] = []
    for activity_type in ACTIVITY_TYPES:
        tp = 0
        fp = 0
        fn = 0
        support = 0
        for p in pairs:
            predicted = p["predicted"]["activity_type"] == activity_type
            actual = p["label"]["activity_type"] == activity_type
            if actual:
                support += 1
            if predicted and actual:
                tp += 1
            elif predicted and not actual:
                fp += 1
            elif not predicted and actual:
                fn += 1
        precision = safe_div(tp, tp + fp)
        recall = safe_div(tp, tp + fn)
        f1 = safe_div(2 * precision * recall, precision + recall)
        results.append(
            {
                "activityType": activity_type,
                "precision": precision,
                "recall": recall,
                "f1": f1,
                "support": support,
            }
        )
    return results


def next_action_date_exact_match_rate(pairs: list[Pair]) -> float:
    if len(pairs) == 0:
        return 0.0
    matches = sum(
        1 for p in pairs if p["predicted"]["next_action_date"] == p["label"]["next_action_date"]
    )
    return matches / len(pairs)


def next_action_null_handling_accuracy(pairs: list[Pair]) -> float:
    if len(pairs) == 0:
        return 0.0
    matches = sum(
        1
        for p in pairs
        if (p["predicted"]["next_action"] is None) == (p["label"]["next_action"] is None)
    )
    return matches / len(pairs)


def _normalize_set_item(item: str) -> str:
    return item.strip().lower()


def _set_precision_recall_f1(predicted: list[str], label: list[str]) -> PrecisionRecallF1:
    pred_set = {_normalize_set_item(item) for item in predicted}
    label_set = {_normalize_set_item(item) for item in label}
    # Both empty is a correct answer, not a miss. See src/lib/metrics.ts for
    # why: scoring it 0 pinned this metric near 0.5 regardless of prompt.
    if len(pred_set) == 0 and len(label_set) == 0:
        return {"precision": 1.0, "recall": 1.0, "f1": 1.0}
    tp = sum(1 for item in pred_set if item in label_set)
    precision = safe_div(tp, len(pred_set))
    recall = safe_div(tp, len(label_set))
    f1 = safe_div(2 * precision * recall, precision + recall)
    return {"precision": precision, "recall": recall, "f1": f1}


def average_set_metrics(pairs: list[Pair], field: SetField) -> PrecisionRecallF1:
    if len(pairs) == 0:
        return {"precision": 0.0, "recall": 0.0, "f1": 0.0}
    per_pair = [_set_precision_recall_f1(p["predicted"][field], p["label"][field]) for p in pairs]

    def avg(key: Literal["precision", "recall", "f1"]) -> float:
        total = 0.0
        for m in per_pair:
            total = total + m[key]
        return total / len(per_pair)

    return {"precision": avg("precision"), "recall": avg("recall"), "f1": avg("f1")}


def hallucination_rate(pairs: list[Pair]) -> float:
    if len(pairs) == 0:
        return 0.0
    hallucinated = 0
    for p in pairs:
        next_action_hallucinated = (
            p["label"]["next_action"] is None and p["predicted"]["next_action"] is not None
        )
        next_action_date_hallucinated = (
            p["label"]["next_action_date"] is None
            and p["predicted"]["next_action_date"] is not None
        )
        if next_action_hallucinated or next_action_date_hallucinated:
            hallucinated += 1
    return hallucinated / len(pairs)
